symbolic-debuginfo/src/sourcebundle/utf8_reader.rs

```diff
@@ -0,0 +1,310 @@
+        let mut next_character_buffer = Vec::from(invalid_portion);
+
+        // read_until_utf8 will not read anything if the buffer is empty,
+        // since an empty buffer is a valid UTF-8 sequence.
+        if self.read_until_utf8(&mut next_character_buffer)? {
+            let next_character_bytes = next_character_buffer.len();
+            let total_bytes_read = valid_up_to + next_character_bytes;
+            buf[valid_up_to..total_bytes_read].copy_from_slice(&next_character_buffer);
+            Ok(total_bytes_read)
+        } else {
+            // We have 4 bytes in the buffer, but they do not form a valid
+            // UTF-8 sequence.
+            Err(Error::new(ErrorKind::InvalidData, UTF8ReaderError::new()))
+        }
+    }
+
+    /// Reads a single byte at a time from the inner reader into the buffer until
+    /// either the buffer contains a valid UTF-8 sequence, the reader is
+    /// exhausted, or the buffer contains 4 bytes (the maximum size of a
+    /// UTF-8 sequence).
+    ///
+    /// If the buffer is empty, we will return Ok(true) without reading anything
+    /// from the reader. An empty byte sequence is a valid UTF-8 sequence (the
+    /// empty string).
+    ///
+    /// Returns whether a valid UTF-8 sequence was found, or an error if the
+    /// reader errors.
+    fn read_until_utf8(&mut self, buffer: &mut Vec<u8>) -> Result<bool> {
+        while std::str::from_utf8(buffer).is_err() {
+            if buffer.len() >= MAX_UTF8_SEQUENCE_SIZE {
+                // We already have 4 bytes in the buffer (maximum UTF-8 sequence size)
+                // so we cannot form a valid UTF-8 sequence by reading more bytes.
+                return Ok(false);
+            }
+
+            let mut byte = [0; 1];
+            if self.inner.read(&mut byte)? == 0 {
+                // Stream has been exhausted without finding
+                // a valid UTF-8 sequence.
+                return Ok(false);
+            }
+
+            buffer.push(byte[0]);
+        }
+
+        Ok(true)
+    }
+}
+
+impl<R> Read for Utf8Reader<R>
+where
+    R: Read,
+{
+    fn read(&mut self, buf: &mut [u8]) -> std::io::Result<usize> {
+        if self.buffer().is_empty() && buf.len() > MAX_BUFFER_SIZE {
+            // If the buffer is bigger than the
+            return self.read_utf8(buf);
+        }
+
+        self.fill_buf()?;
+
+        let bytes_to_copy = std::cmp::min(buf.len(), self.buffer().len());
+        buf[..bytes_to_copy].copy_from_slice(&self.buffer()[..bytes_to_copy]);
+        self.consume(bytes_to_copy);
+
+        Ok(bytes_to_copy)
+    }
+}
+
+impl<R> BufRead for Utf8Reader<R>
+where
+    R: Read,
+{
+    fn fill_buf(&mut self) -> std::io::Result<&[u8]> {
+        if !self.buffer().is_empty() {
+            return Ok(self.buffer());
+        }
+
+        let mut buf = [0; MAX_BUFFER_SIZE];
+        let bytes_read = self.read_utf8(&mut buf)?;
+
+        self.buffer_vec = buf[..bytes_read].into();
```

Comment:
You can reuse the allocation here.

```suggestion
        self.buffer_vec.clear();
        self.buffer_vec.extend_from_slice(&buf[..bytes_read]);
```

Reply:
I think this still copies, which @swatinem asked me to avoid doing. Perhaps we can combine the two suggestions